File: frontend/src/types/contracts.ts
export type MotionState = 'stationary' | 'walking_slow' | 'walking_fast' | 'running';
export type DistanceCategory = 'very_close' | 'mid' | 'far';
export type NavigationMode = 'NAVIGATION' | 'EXPLORE' | 'READ' | 'QUIET';
export type CarryMode = 'hand_held' | 'necklace' | 'chest_clip' | 'pocket';
export type CognitionLayer = 'l1_survival' | 'l2_edge' | 'l3_cloud';
export type HumanHelpProvider = 'twilio' | 'livekit';

export const HAZARD_TYPE = {
  UNSPECIFIED: 0,
  DROP_AHEAD: 1,
  STATIC_OBSTACLE: 2,
  DYNAMIC_OBSTACLE: 3,
  VEHICLE: 4,
} as const;

export type HazardType = typeof HAZARD_TYPE[keyof typeof HAZARD_TYPE];

export interface MotionSnapshot {
  state: MotionState;
  pitch: number;
  velocity: number;
}

export interface MultimodalFrameMessage {
  type: 'multimodal_frame';
  session_id: string;
  timestamp_ms: number;
  frame_jpeg_base64?: string;
  motion_state: MotionState;
  pitch: number;
  velocity: number;
  user_text?: string;
  /** Góc xoay ngang tích lũy (độ) kể từ frame trước → Semantic Odometry */
  yaw_delta_deg?: number;
  carry_mode?: CarryMode;
  sensor_unavailable?: boolean;
  lat?: number;
  lng?: number;
  heading_deg?: number;
  location_accuracy_m?: number;
  location_age_ms?: number;
  edge_semantic_cues: EdgeSemanticCueMessage[];
}

export interface AudioChunkMessage {
  type: 'audio_chunk';
  session_id: string;
  timestamp_ms: number;
  audio_chunk_pcm16: string;
}

export interface UserCommandMessage {
  type: 'user_command';
  session_id: string;
  timestamp_ms: number;
  command: string;
}

export interface EdgeSemanticCueMessage {
  cue_type: string;
  text?: string;
  confidence: number;
  position_x?: number;
  distance_m?: number;
  position_clock?: string;
  ttl_ms?: number;
  source: string;
}

export interface HazardObservationMessage {
  type: 'hazard_observation';
  session_id: string;
  timestamp_ms: number;
  hazard_type: HazardType;
  bearing_x?: number;
  distance_m: number;
  relative_velocity_mps: number;
  confidence?: number;
  source?: string;
  suppress_ms?: number;
}

export interface AssistantTextMessage {
  type: 'assistant_text';
  session_id: string;
  timestamp_ms: number;
  text: string;
}

export interface AssistantAudioMessage {
  type: 'audio_chunk';
  session_id: string;
  timestamp_ms: number;
  pcm24?: string;
  pcm16?: string;
  hazard_position_x?: number;
}

export interface SemanticCueMessage {
  type: 'semantic_cue';
  cue: 'approaching_object' | 'soft_obstacle' | 'turning_recommended' | 'destination_near' | 'pocket_mode_active';
  position_x?: number;
}

export interface HardStopMessage {
  type: 'HARD_STOP';
  position_x: number;
  distance: DistanceCategory;
  hazard_type: string;
  confidence: number;
  ts?: string;
}

export interface ConnectionStateMessage {
  type: 'connection_state';
  state: 'connected' | 'reconnecting' | 'disconnected' | 'degraded';
  detail?: string;
}

export interface SafetyDirectiveMessage {
  type: 'safety_directive';
  session_id: string;
  timestamp_ms: number;
  hazard_type?: HazardType;
  hazard_score: number;
  hard_stop: boolean;
  haptic_intensity: number;
  spatial_audio_pitch_hz: number;
  spatial_audio_pan: number;
  needs_human_assistance: boolean;
  reason?: string;
  flush_audio: boolean;
}

export interface HumanHelpSessionMessage {
  type: 'human_help_session';
  session_id: string;
  timestamp_ms: number;
  help_link?: string;
  rtc: {
    provider: HumanHelpProvider;
    room_name: string;
    identity?: string;
    token: string;
    expires_in: number;
  };
}

export interface CognitionStateMessage {
  type: 'cognition_state';
  session_id: string;
  timestamp_ms: number;
  active_layer: CognitionLayer;
  cloud_link_healthy: boolean;
  edge_cognition_available: boolean;
  cloud_rtt_ms?: number;
  reason?: string;
}

export type BackendToClientMessage =
  | AssistantTextMessage
  | AssistantAudioMessage
  | SafetyDirectiveMessage
  | ConnectionStateMessage
  | SemanticCueMessage
  | HumanHelpSessionMessage
  | CognitionStateMessage;

export function hazardTypeToLabel(hazardType?: HazardType | null): string {
  switch (hazardType) {
    case HAZARD_TYPE.DROP_AHEAD:
      return 'drop_ahead';
    case HAZARD_TYPE.STATIC_OBSTACLE:
      return 'static_obstacle';
    case HAZARD_TYPE.DYNAMIC_OBSTACLE:
      return 'dynamic_obstacle';
    case HAZARD_TYPE.VEHICLE:
      return 'vehicle';
    default:
      return 'unspecified';
  }
}

export function parseHazardType(value: string | HazardType | null | undefined): HazardType {
  if (typeof value === 'number') {
    if (Object.values(HAZARD_TYPE).includes(value as HazardType)) {
      return value as HazardType;
    }
    return HAZARD_TYPE.UNSPECIFIED;
  }

  const normalized = value?.trim().toLowerCase();
  switch (normalized) {
    case 'drop_ahead':
    case 'dropahead':
    case 'edge_drop_hazard':
      return HAZARD_TYPE.DROP_AHEAD;
    case 'static_obstacle':
    case 'staticobstacle':
    case 'pole':
      return HAZARD_TYPE.STATIC_OBSTACLE;
    case 'dynamic_obstacle':
    case 'dynamicobstacle':
    case 'moving_obstacle':
      return HAZARD_TYPE.DYNAMIC_OBSTACLE;
    case 'vehicle':
    case 'bike':
    case 'motorbike':
    case 'car':
      return HAZARD_TYPE.VEHICLE;
    default:
      return HAZARD_TYPE.UNSPECIFIED;
  }
}

export function distanceCategoryToMeters(distance: DistanceCategory): number {
  switch (distance) {
    case 'very_close':
      return 0.45;
    case 'mid':
      return 1.5;
    case 'far':
      return 3.5;
  }
}

export function metersToDistanceCategory(distanceM: number): DistanceCategory {
  if (distanceM <= 0.8) {
    return 'very_close';
  }
  if (distanceM <= 2.5) {
    return 'mid';
  }
  return 'far';
}

File: crates/apollos-proto/src/contracts.rs
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MotionState {
    Stationary,
    WalkingSlow,
    WalkingFast,
    Running,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DistanceCategory {
    VeryClose,
    Mid,
    Far,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize_repr, Deserialize_repr)]
#[repr(i32)]
pub enum HazardType {
    Unspecified = 0,
    DropAhead = 1,
    StaticObstacle = 2,
    DynamicObstacle = 3,
    Vehicle = 4,
}

impl HazardType {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Unspecified => "unspecified",
            Self::DropAhead => "drop_ahead",
            Self::StaticObstacle => "static_obstacle",
            Self::DynamicObstacle => "dynamic_obstacle",
            Self::Vehicle => "vehicle",
        }
    }

    pub fn from_str(value: &str) -> Option<Self> {
        let normalized = value.trim().to_ascii_lowercase();
        match normalized.as_str() {
            "drop_ahead" | "dropahead" | "edge_drop_hazard" => Some(Self::DropAhead),
            "static_obstacle" | "staticobstacle" | "pole" => Some(Self::StaticObstacle),
            "dynamic_obstacle" | "dynamicobstacle" | "moving_obstacle" => {
                Some(Self::DynamicObstacle)
            }
            "vehicle" | "bike" | "motorbike" | "car" => Some(Self::Vehicle),
            "unspecified" | "unknown_hazard" | "unknown" => Some(Self::Unspecified),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum NavigationMode {
    #[serde(rename = "NAVIGATION")]
    Navigation,
    #[serde(rename = "EXPLORE")]
    Explore,
    #[serde(rename = "READ")]
    Read,
    #[serde(rename = "QUIET")]
    Quiet,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CarryMode {
    HandHeld,
    Necklace,
    ChestClip,
    Pocket,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CognitionLayer {
    L1Survival,
    L2Edge,
    L3Cloud,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MotionSnapshot {
    pub state: MotionState,
    pub pitch: f32,
    pub velocity: f32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SensorHealthSnapshot {
    pub score: f32,
    pub flags: Vec<String>,
    pub degraded: bool,
    pub source: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SensorUncertaintySnapshot {
    pub covariance_3x3: Vec<f32>,
    pub innovation_norm: f32,
    pub source: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CloudLinkSnapshot {
    pub connected: bool,
    pub rtt_ms: Option<f32>,
    pub source: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct VisionOdometrySnapshot {
    pub source: String,
    pub applied: bool,
    pub optical_flow_score: Option<f32>,
    pub variance_m2: Option<f32>,
    pub pose_x_m: Option<f32>,
    pub pose_y_m: Option<f32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EdgeSemanticCueMessage {
    pub cue_type: String,
    pub text: Option<String>,
    pub confidence: f32,
    pub position_x: Option<f32>,
    pub distance_m: Option<f32>,
    pub position_clock: Option<String>,
    pub ttl_ms: Option<u32>,
    pub source: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MultimodalFrameMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub frame_jpeg_base64: Option<String>,
    pub motion_state: MotionState,
    pub pitch: f32,
    pub velocity: f32,
    pub user_text: Option<String>,
    pub yaw_delta_deg: Option<f32>,
    pub carry_mode: Option<CarryMode>,
    pub sensor_unavailable: Option<bool>,
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub heading_deg: Option<f32>,
    pub location_accuracy_m: Option<f32>,
    pub location_age_ms: Option<u64>,
    pub sensor_health: Option<SensorHealthSnapshot>,
    pub sensor_uncertainty: Option<SensorUncertaintySnapshot>,
    pub vision_odometry: Option<VisionOdometrySnapshot>,
    pub cloud_link: Option<CloudLinkSnapshot>,
    pub edge_semantic_cues: Vec<EdgeSemanticCueMessage>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AudioChunkMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub audio_chunk_pcm16: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UserCommandMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub command: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HazardObservationMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub hazard_type: HazardType,
    pub bearing_x: Option<f32>,
    pub distance_m: f32,
    pub relative_velocity_mps: f32,
    pub confidence: Option<f32>,
    pub source: Option<String>,
    pub suppress_ms: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AssistantTextMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AssistantAudioMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub pcm24: Option<String>,
    pub pcm16: Option<String>,
    pub hazard_position_x: Option<f32>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SemanticCue {
    ApproachingObject,
    SoftObstacle,
    TurningRecommended,
    DestinationNear,
    PocketModeActive,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SemanticCueMessage {
    pub cue: SemanticCue,
    pub position_x: Option<f32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SafetyDirectiveMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub hazard_type: Option<HazardType>,
    pub hazard_score: f32,
    pub hard_stop: bool,
    pub haptic_intensity: f32,
    pub spatial_audio_pitch_hz: f32,
    pub spatial_audio_pan: f32,
    pub needs_human_assistance: bool,
    pub reason: Option<String>,
    pub flush_audio: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionState {
    Connected,
    Reconnecting,
    Disconnected,
    Degraded,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ConnectionStateMessage {
    pub state: ConnectionState,
    pub detail: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CognitionStateMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub active_layer: CognitionLayer,
    pub cloud_link_healthy: bool,
    pub edge_cognition_available: bool,
    pub cloud_rtt_ms: Option<f32>,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HumanHelpProvider {
    Twilio,
    Livekit,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HumanHelpRtcSession {
    pub provider: HumanHelpProvider,
    pub room_name: String,
    pub identity: Option<String>,
    pub token: String,
    pub expires_in: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HumanHelpSessionMessage {
    pub session_id: String,
    pub timestamp_ms: u64,
    pub help_link: Option<String>,
    pub rtc: HumanHelpRtcSession,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientToBackendMessage {
    #[serde(rename = "multimodal_frame")]
    MultimodalFrame(MultimodalFrameMessage),
    #[serde(rename = "audio_chunk")]
    AudioChunk(AudioChunkMessage),
    #[serde(rename = "user_command")]
    UserCommand(UserCommandMessage),
    #[serde(rename = "hazard_observation")]
    HazardObservation(HazardObservationMessage),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum BackendToClientMessage {
    #[serde(rename = "assistant_text")]
    AssistantText(AssistantTextMessage),
    #[serde(rename = "audio_chunk")]
    AssistantAudio(AssistantAudioMessage),
    #[serde(rename = "safety_directive")]
    SafetyDirective(SafetyDirectiveMessage),
    #[serde(rename = "connection_state")]
    ConnectionState(ConnectionStateMessage),
    #[serde(rename = "semantic_cue")]
    SemanticCue(SemanticCueMessage),
    #[serde(rename = "human_help_session")]
    HumanHelpSession(HumanHelpSessionMessage),
    #[serde(rename = "cognition_state")]
    CognitionState(CognitionStateMessage),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serializes_multimodal_frame_with_type_tag() {
        let payload = ClientToBackendMessage::MultimodalFrame(MultimodalFrameMessage {
            session_id: "s1".to_string(),
            timestamp_ms: 1_741_169_200_000,
            frame_jpeg_base64: None,
            motion_state: MotionState::WalkingFast,
            pitch: 10.0,
            velocity: 2.2,
            user_text: None,
            yaw_delta_deg: Some(6.0),
            carry_mode: Some(CarryMode::Necklace),
            sensor_unavailable: Some(false),
            lat: None,
            lng: None,
            heading_deg: None,
            location_accuracy_m: None,
            location_age_ms: None,
            sensor_health: None,
            sensor_uncertainty: None,
            vision_odometry: None,
            cloud_link: None,
            edge_semantic_cues: Vec::new(),
        });

        let json = serde_json::to_string(&payload).expect("should serialize");
        assert!(json.contains("\"type\":\"multimodal_frame\""));
        assert!(json.contains("\"motion_state\":\"walking_fast\""));
    }
}
